src/huggingface_hub/commit_api.py

```diff
@@ -0,0 +1,424 @@
+"""
+Type definitions and utilities for the `create_commit` API
+"""
+
+import base64
+import io
+import os
+import sys
+from concurrent import futures
+from concurrent.futures import ThreadPoolExecutor
+from contextlib import contextmanager
+from dataclasses import dataclass, field
+from typing import BinaryIO, Dict, Iterable, List, Optional, Tuple, Union
+
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+import requests
+
+from .constants import ENDPOINT
+from .lfs import UploadInfo, lfs_upload, post_lfs_batch_info, validate_batch_actions
+from .utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+UploadMode = Literal["lfs", "regular"]
+
+
+@dataclass
+class CommitOperationDelete:
+    """
+    Data structure holding necessary info to delete
+    a file from a repository on the Hub
+
+    Args:
+        path_in_repo (`str`):
+            Relative filepath in the repo, for example:
+            `"checkpoints/1fec34a/weights.bin"`
+    """
+
+    path_in_repo: str
+
+
+@dataclass
+class CommitOperationAdd:
+    """
+    Data structure holding necessary info to upload a file
+    to a repository on the Hub
+
+    Args:
+        path_in_repo (`str`):
+            Relative filepath in the repo, for example:
+            `"checkpoints/1fec34a/weights.bin"`
+        path_or_fileobj (`str`, `bytes`, or `BinaryIO`):
+            Either:
+            - a path to a local file (as str) to upload
+            - a buffer of bytes (`bytes`) holding the content of the file to upload
+            - a "file object" (subclass of `io.BufferedIOBase`), typically obtained
+                with `open(path, "rb")`
+    """
+
+    path_in_repo: str
+    path_or_fileobj: Union[str, bytes, BinaryIO]
+
+    _upload_info: Optional[UploadInfo] = field(default=None, init=False)
+
+    def validate(self):
+        """
+        Ensures `path_or_fileobj` is valid.
```

Comment:
Could you please also explain what a valid `path_or_fileobj` is?

Also, I wonder why we don't use the `file` pattern which is common in python, to be a file object or a path anyway, while having a much shorter parameter name.

Reply:
> I wonder why we don't use the file pattern

Usually, a `file` argument is expected to have a `write` method, no? See https://docs.python.org/fr/3/library/functions.html#print for example

Here we also accept a file path (`str`), hence `path_or_fileobj`

Reply:
From a personal POV, I like the explicitness of `path_or_fileobj`, which is clearly either a path or a [file object](https://docs.python.org/3/glossary.html#term-file-object)